xclim/sdba/_adjustment.py

```diff
@@ -944,3 +944,378 @@ def extremes_adjust(
+        _otc_adjust,
+        hist,
+        ref,
+        kwargs=dict(
+            bin_width=bin_width,
+            bin_origin=bin_origin,
+            num_iter_max=num_iter_max,
+            jitter_inside_bins=jitter_inside_bins,
+        ),
+        input_core_dims=[["dim_hist", pts_dim], ["dim_ref", pts_dim]],
+        output_core_dims=[["dim_hist", pts_dim]],
+        keep_attrs=True,
+        vectorize=True,
+    )
+
+    # Pad dim differences with NA to please map_blocks
+    ref = ref.unstack().rename({v: k for k, v in ref_map.items()})
+    scen = scen.unstack().rename("scen")
+    for d in dim:
+        full_d = xr.concat([ref[d], scen[d]], dim=d).drop_duplicates(d)
+        scen = scen.reindex({d: full_d})
+
+    return scen.to_dataset()
+
+
+def _dotc_adjust(
+    X1: np.ndarray,
+    Y0: np.ndarray,
+    X0: np.ndarray,
+    bin_width: list | None = None,
+    bin_origin: list | None = None,
+    num_iter_max: int | None = 100_000_000,
+    cov_factor: str = "std",
+    jitter_inside_bins: bool = True,
+    kind: dict | None = None,
+):
+    """Dynamical Optimal Transport Correction of the bias of X with respect to Y.
+
+    Parameters
+    ----------
+    X1 : np.ndarray
+        Simulation data to adjust.
+    Y0 : np.ndarray
+        Bias correction reference.
+    X0 : np.ndarray
+        Historical simulation data.
+    bin_width : list | None
+        Bin widths for all dimensions.
+    bin_origin : list | None
+        Bin origins for all dimensions.
+    num_iter_max : int | None
+        Maximum number of iterations used in the earth mover distance algorithm.
+    cov_factor : str = "std"
+        Rescaling factor.
+
+    Returns
+    -------
+    np.ndarray
+        Adjusted data
+
+    References
+    ----------
+    :cite:cts:`sdba-robin_2021`
+    """
+    # Initialize parameters
+    if bin_width is None:
+        bin_width = u.bin_width_estimator([Y0, X0, X1])
+    elif isinstance(bin_width, list):
+        bin_width = np.array(bin_width)
+
+    # Map ref to hist
+    yX0 = _otc_adjust(
+        Y0,
+        X0,
+        bin_width=bin_width,
+        bin_origin=bin_origin,
+        num_iter_max=num_iter_max,
+        jitter_inside_bins=False,
+    )
+
+    # Map hist to sim
+    yX1 = _otc_adjust(
+        yX0,
+        X1,
+        bin_width=bin_width,
+        bin_origin=bin_origin,
+        num_iter_max=num_iter_max,
+        jitter_inside_bins=False,
+    )
+
+    # Temporal evolution
+    motion = np.empty(yX0.shape)
+    for j in range(yX0.shape[1]):
+        if kind is not None and j in kind.keys() and kind[j] == "*":
+            motion[:, j] = yX1[:, j] / yX0[:, j]
+        else:
+            motion[:, j] = yX1[:, j] - yX0[:, j]
+
+    # Apply a variance dependent rescaling factor
+    if cov_factor == "cholesky":
+        fact0 = u.eps_cholesky(np.cov(Y0, rowvar=False))
+        fact1 = u.eps_cholesky(np.cov(X0, rowvar=False))
+        motion = motion @ fact0 @ np.linalg.inv(fact1)
```

Comment:
```suggestion
        motion = motion @ np.linalg.inv(fact1) @ fact0
```
no? matrices don't commute generally, and I believe we first need to remove the standard deviation, then add the one from the reference

Reply:
Sidebar: What am I looking at here?? What does `@` do??

Reply:
[matrix multiplication shorthand in python](https://peps.python.org/pep-0465/)

Reply:
`@` is matrix multiplication. I thought it is simpler than SBCK's `apply_along_axis` but maybe not.
`fact0 @ np.linalg.inv(fact1) @ motion` is probably the best way to write it.

Reply:
yes that's another point, I find this tricky. `fact0, fact1` have shape `len(multivar), len(multivar)`, but they're not symmetric, is that right? so `fact0 @ np.linalg.inv(fact1) @ motion.T` and `fact0.T @ np.linalg.inv(fact1).T @ motion.T` will both work but only one is correct